src/macaron/repo_finder/provenance_extractor.py

```diff
@@ -355,3 +358,354 @@ def check_if_repository_purl_and_url_match(url: str, repo_purl: PackageURL) -> b
         purl_path = f"{repo_purl.namespace}/{purl_path}"
     # Note that the urllib method includes the "/" before path while the PURL method does not.
     return f"{parsed_url.hostname}{parsed_url.path}".lower() == f"{expanded_purl_type or repo_purl.type}/{purl_path}"
+
+
+class ProvenanceBuildDefinition(ABC):
+    """Abstract base class for representing provenance build definitions.
+
+    This class serves as a blueprint for various types of build definitions
+    in provenance data. It outlines the methods and properties that derived
+    classes must implement to handle specific build definition types.
+    """
+
+    #: Determines the expected ``buildType`` field in the provenance predicate.
+    expected_build_type: str
+
+    @abstractmethod
+    def get_build_invocation(self, statement: InTotoV01Statement | InTotoV1Statement) -> tuple[str | None, str | None]:
+        """Retrieve the build invocation information from the given statement.
+
+        This method is intended to be implemented by subclasses to extract
+        specific invocation details from a provenance statement.
+
+        Parameters
+        ----------
+        statement : InTotoV1Statement | InTotoV01Statement
+            The provenance statement from which to extract the build invocation
+            details. This statement contains the metadata about the build process
+            and its associated artifacts.
+
+        Returns
+        -------
+        tuple[str | None, str | None]
+            A tuple containing two elements:
+            - The first element is the build invocation entry point (e.g., workflow name), or None if not found.
+            - The second element is the invocation URL or identifier (e.g., job URL), or None if not found.
+
+        Raises
+        ------
+        NotImplementedError
+            If the method is called directly without being overridden in a subclass.
+        """
+
+
+class SLSAGithubGenericBuildDefinitionV01(ProvenanceBuildDefinition):
+    """Class representing the SLSA GitHub Generic Build Definition (v0.1).
+
+    This class implements the abstract methods defined in `ProvenanceBuildDefinition`
+    to extract build invocation details specific to the GitHub provenance generator's generic build type.
+    """
+
+    #: Determines the expected ``buildType`` field in the provenance predicate.
+    expected_build_type = "https://github.com/slsa-framework/slsa-github-generator/generic@v1"
+
+    def get_build_invocation(self, statement: InTotoV01Statement | InTotoV1Statement) -> tuple[str | None, str | None]:
+        """Retrieve the build invocation information from the given statement.
+
+        This method is intended to be implemented by subclasses to extract
+        specific invocation details from a provenance statement.
```

Comment:
Should we remove these 2 lines as this is a concrete method 🤔 ?
The same comment applies for `SLSAGithubActionsBuildDefinitionV1.get_build_invocation`

Reply:
See commit [17f3453](https://github.com/oracle/macaron/pull/872/commits/17f3453c4be75784f00a422b71ea829e3f04e4de).